src/main/resources/static/js/posts.js

```diff
@@ -0,0 +1,305 @@
+// TODO:
+// 1. Message ID we sent needs to match response
+// 2. Origin of the response should match the oidc auth URI origin
+// 3. verify `data` response is an object
+// 4. verify data.subject = lti.put_data.response
+
+// https://www.imsglobal.org/spec/lti-cs-oidc/v0p1#js-example-listen-for-put_data-response
+
+//  -------------- UTILITY FUNCTIONS --------------- //
+
+// Simple UUID generator
+const create_UUID = () => {
+  var dt = new Date().getTime();
+  var uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
+    /[xy]/g,
+    function (c) {
+      var r = (dt + Math.random() * 16) % 16 | 0;
+      dt = Math.floor(dt / 16);
+      return (c == "x" ? r : (r & 0x3) | 0x8).toString(16);
+    }
+  );
+  return uuid;
+};
+
+//  -------------- CONSTANTS AND PLACEHOLDERS --------------- //
+
+// Defined here so we only have to change them in one place
+const PUT_SUBJECT = "lti.put_data";
+const GET_SUBJECT = "lti.get_data";
+
+// Generate a UUID for use in sending and verifying responses
+const uuid = create_UUID();
+
+// Flag to auto-redirect
+const autoRedirect = true;
+// redirectUrl if we want to use it
+let redirectUrl = null;
+
+// placeholder for plaftormOrigin
+let platformOrigin = null;
+
+// placeholder for mesage type sent
+let messageType = null;
+
+// placeholder for expected state and nonce
+let expected_state = null;
+let expected_nonce = null;
+
+// placeholder for successful 'get' responses
+let stateResponseData = null;
+let nonceResponseData = null;
+
+//  -------------- APP LOGIC --------------- //
+
+// Call from the template passing the url for the next step
+// set autoRedirect=true above if we want to proceed to the next step after responses and validation
+const setAutoRedirectUrl = (oidcEndpointComplete) => {
+  redirectUrl = oidcEndpointComplete;
+};
+
+// Validation function before sending postMessage requests
+const validateSendConditions = (state, nonce, target) => {
+  if (state && nonce && target && platformOrigin) {
+    console.log("✅ Passed send conditions");
+    return true;
+  }
+  return false;
+};
+
+const validateResponseConditions = (origin, data) => {
+  // Conditions in the spec
+  // https://www.imsglobal.org/spec/lti-cs-oidc/v0p1#js-example-listen-for-put_data-response
+
+  // should this be a get or put response subject?
+  const subjectResponse =
+    messageType === "get"
+      ? `${GET_SUBJECT}.response`
+      : `${PUT_SUBJECT}.response`;
+
+  if (
+    // verify `data` is an object
+    typeof data === "object" &&
+    // Message ID we sent needs to match response
+    data.message_id === uuid &&
+    // Origin of the response should match the oidc auth URI origin
+//    origin === platformOrigin &&
```

Comment:
Was this commented out because of Canvas? Is Canvas the only offender here? I think we should at least explain why this is commented out if not set something up to only skip this validation for Canvas or something like that.